Prevent browser caching during auth flow

johejo opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

Add some headers to prevent browser caching during authentication flow.

<!--- Describe your changes in detail -->

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->

<!--- If it fixes an open issue, please link to the issue here. -->

Browser might cache `/oauth2/start` request etc. under certain conditions.

In my environment, Chrome will **not**  send `GET /oauth2/start` to the oauth2_proxy,  and the Chrome re-uses previous `GET /authorize?approval_prompt=force&client_id=xxx&redirect_uri=yyyy.......` request and send it to the provider.

Below are my steps for browser to reproduce.
1. Login as usual our site using oauth2_proxy. ( and open a blank page on the next tab for step 4 )
2. Clear all cookies (for our site and provider)
3. Close our site tab
4. Reopen tab (Right click on the top of the empty tab)
5. Login with Provider
6. oauth2_proxy returns 403

I think oauth2_proxy gets bad OIDC state param.
If my explanation is not enough, I would supplement more further.

The implementation using some headers to prevent caching referred from go-chi/chi.
https://github.com/go-chi/chi/blob/master/middleware/nocache.go

## How Has This Been Tested?

Added unit tests.

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

JoelSpeed commented:
I haven't reviewed the code, but the motivation seems good. I'd prefer to see links to official docs from the browser providers rather than referring to another project though. Any chance you can link some docs in this PR and within the comments in the code as well?

johejo commented:
Sorry, I couldn't find official announcement from the browser provider.
Exploring the chromium source code, but that was difficult for me.

Google and w3c documentation says belows.
https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/http-caching?hl=en
https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.9

johejo commented:
A related issue was opened in ingress-nginx by my colleague.
https://github.com/kubernetes/ingress-nginx/issues/5286

johejo commented:
Hi, thank you for your input.
I've fixed the header addition logic a bit and add tests, referring to Google's documentation and such.
And I've updated changelog.

johejo commented:
I added test. 
https://github.com/oauth2-proxy/oauth2-proxy/pull/453/commits/2dca7fbc2cc4ee9a46c9ed58dbdca02214bfb9c0

ploxiln commented on oauthproxy.go:
```diff
@@ -633,7 +633,30 @@ func getRemoteAddr(req *http.Request) (s string) {
+// See https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/http-caching?hl=en
+var noCacheHeaders = map[string]string{
+	"Expires":         time.Unix(0, 0).Format(time.RFC1123),
+	"Cache-Control":   "no-cache, no-store, must-revalidate, max-age=0",
```
it seems like the only header needed should be `Cache-Control: no-store` and this is confirmed by https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control

johejo replied:
Your point is correct.
In reality, however, many servers set multiple values to cache-control to prevent caching.

I've run the chrome dev tool on the login screen of many web services (google, microsoft, etc.) to check the headers.
Many of them were `cache-control: no-cache, no-store, max-age=0, must-revalidate`.
Why is that?

ploxiln replied:
"copy-paste and it seems to always work so far" is as likely as any other explanation. But fair enough, it'll work.

ploxiln commented on oauthproxy.go:
```diff
@@ -633,7 +633,30 @@ func getRemoteAddr(req *http.Request) (s string) {
+}
+
+func hasProxyPrefix(path, prefix string) bool {
+	return strings.HasPrefix(path, prefix)
```
`hasProxyPrefix()` is exactly equivalent to `strings.HasPrefix()`, so it seems extraneous to me

ploxiln commented on oauthproxy.go:
```diff
@@ -633,7 +633,30 @@ func getRemoteAddr(req *http.Request) (s string) {
+
 func (p *OAuthProxy) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	if hasProxyPrefix(req.URL.Path, p.ProxyPrefix) {
+		prepareNoCache(rw)
```
In "proxy" mode (in contrast to "auth_request" mode), oauth2-proxy serves the sign-in page directly when you try to access some other url and are not authenticated, or it serves the redirect to the provider to start the oauth process. These responses are to the application URL and do not have the proxy prefix, but we also want to prevent the browser caching these. (Even if oauth2-proxy were to serve a redirect to its own path first, we would still want to prevent that redirect from being cached.)

So, a possible fix would be to add these headers at the start of the `OAuthProxy.SignInPage()` and `OAuthProxy.OAuthStart()` methods. There might be more thorough or elegant strategies, but I'm not sure ...

johejo replied:
Normally, we should use http middleware to do those process, but the structure of oauth2-proxy's http handler is a bit complicated, so it's difficult.
For now, I added prepareNoCache processing to SignInPage and OAuthStart.

ploxiln replied:
looks good to me
